### PR TITLE
Upgrade SvelteKit

### LINKED ISSUE #55: Decrease site navigation loading time by fixing prefetching with SvelteKit upgrade
**Describe the bug**
Clicking links to other pages in the navbar is slower when on desktop since prefetching doesn't work in the current pinned SvelteKit version (it works on mobile, just not desktop).

Version in use: [1.0.0-next.202](https://github.com/hack4impact-uiuc/website/blob/2df77c5ddb2436f78b42b0c4b7c6fac4b3b65838/package.json#L24)

*SvelteKit changelog*:
`sveltekit:prefetch` breaks at: [1.0.0-next.199](https://github.com/sveltejs/kit/blob/master/packages/kit/CHANGELOG.md#100-next199)
- Reason: New method of finding `a` tags uses some obscure function that doesn't work in a timeout.

`sveltekit:prefetch` fixed at: [1.0.0-next.203](https://github.com/sveltejs/kit/blob/master/packages/kit/CHANGELOG.md#100-next203)
- Reason: They used an event to indirectly read the node after the timeout. See comment: https://github.com/sveltejs/kit/blob/f196861d8174a1bfcf81778090311aea9f8d6535/packages/kit/src/runtime/client/client.js#L1038-L1039

**To Reproduce**

1. Go to [https://uiuc.hack4impact.org/](https://uiuc.hack4impact.org/)
2. Open up the network tab in devtools
3. Hover over an `a` tag with sveltekit:prefetch
4. Notice 0 network requests are made, when it *should* load the JS chunks for such page

**Expected behavior**
Chunks are loaded on hover for prefetched pages.

I suggest upgrading to [1.0.0-next.203](https://github.com/sveltejs/kit/blob/master/packages/kit/CHANGELOG.md#100-next203).

### LINKED ISSUE #56: Reduce route loading latency by preventing refetching of layout.
**Describe the the feature request**
At the moment, the layout component accesses the page's path value from SvelteKits parameter in the load function. When this value is accessed, SvelteKit reruns the corresponding load function whenever the value changes (a page navigation resulting in a new path). This means the layout's load function runs first, sequentially, before any page load functions, on every navigation, causing an additional 100ms delay before page data is loaded (since the layout load function queries the API).

I think this is unnecessary since the layout doesn't really change from anything to do with loading per se, it changes because the URL path changes. The URL path can be effectively acquired reactively in other manners instead of the layout's load method. This would prevent the refetch on every page navigation.

Potential ways to get the URL path without using the load function's parameter:
1. I believe the sveltekit navigation end event could cover this, although the docs deleted it so I'd have to check. Coincidentally it's already being used in the Nav component which is just where the URL path is needed.
2. The page store from SvelteKit's $app/stores package. Pretty simple, and it automatically reacts.

### PR DESCRIPTION
**Pull request by AndrewLester**
## Status:
:rocket: Ready

## Description

Upgrades SvelteKit version to fix prefetching.

Fixes: #55 
Fixes: #56 

## Screenshots

Ignore the errors, I can't connect to the API.

https://user-images.githubusercontent.com/23221268/158862741-293d1fc6-3be3-4b5d-aea4-f6e98c166d41.mov

### CODE REVIEW

**Review by SirajChokshi — CHANGES_REQUESTED** [review]
Thanks for maintaining this.

The pre-fetching is great, but it seems like the mobile navigation toggle is broken (https://deploy-preview-57--h4iuiuc.netlify.app/)

**Comment by AndrewLester** [comment]
> Thanks for maintaining this.
> 
> The pre-fetching is great, but it seems like the mobile navigation toggle is broken (https://deploy-preview-57--h4iuiuc.netlify.app/)

That's unfortunate, the preview deployments really come in handy. After checking the console, I think some script is erroring out and it's crashing Svelte before it can register the button on:click listeners. Here's a screenshot of the error's location in the script, looks like it might be ackee?
image

Here's a screenshot of a potential root cause (note the `{}.CLIENT_ACKEE_URL`, which yields undefined):
image


You might have a better idea of why this is happening, I'm not entirely sure though. Maybe the Ackee urls aren't in the preview?

**Comment by arpanlaha** [comment]
> > Thanks for maintaining this.
> > The pre-fetching is great, but it seems like the mobile navigation toggle is broken (https://deploy-preview-57--h4iuiuc.netlify.app/)
> 
> That's unfortunate, the preview deployments really come in handy. After checking the console, I think some script is erroring out and it's crashing Svelte before it can register the button on:click listeners. Here's a screenshot of the error's location in the script, looks like it might be ackee? image
> 
> Here's a screenshot of a potential root cause (note the `{}.CLIENT_ACKEE_URL`, which yields undefined): image
> 
> You might have a better idea of why this is happening, I'm not entirely sure though. Maybe the Ackee urls aren't in the preview?

Ackee was me, I'll take a look soon (probably the weekend).

**Comment by AndrewLester** [comment]
Coming back to this, it might just be better to upgrade SvelteKit to its most recent version since a lot has changed. The SvelteKit docs really only describe its most recent version, so it will be easier to maintain with a newer version in the codebase. I've recently done a SvelteKit upgrade, so I know generally what needs to be changed in the codebase and I could handle it in this PR. Thoughts?

**Comment by arpanlaha** [comment]
> Coming back to this, it might just be better to upgrade SvelteKit to its most recent version since a lot has changed. The SvelteKit docs really only describe its most recent version, so it will be easier to maintain with a newer version in the codebase. I've recently done a SvelteKit upgrade, so I know generally what needs to be changed in the codebase and I could handle it in this PR. Thoughts?

Sure go for it, sorry I forgot to follow up earlier. Let me know once you've bumped to a more up-to-date version and I'll take a look again.

**Comment by AndrewLester** [comment]
This PR unfortunately became quite large because of the upgrade. Here's a rundown of why I made some changes that increased the files changed by a lot:
1. Removing the custom import alias for `$lib` (was pointing at `src`). I removed this because it was messing up SvelteKit's new generated types system. In doing so, I added new import aliases which resulted in almost every one changing.
2. More explicit typing of endpoint and page load functions. Explicitly typing these functions allows SvelteKit's generated types to work when necessary, meaning page params are typed. I wanted to make this consistent so I typed every one, changing all of these functions to arrow functions in the process. Since not all of them made use of SvelteKit's generated types, for some I just imported their type from the normal SvelteKit type.
3. Changed type definitions from `interface` to `type`. I made this change because of an issue I came across before when returning `interface` defined types in endpoint functions. For some reason, SvelteKit's type system doesn't think `interface` defined types are JSON compatible. Swapping to `type` fixes this issue. [This TS playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAUgygeQHIDUCGAbArtAvFAZ2ACcBLAOwHMoAfWRJBAIwCsIBjYAbgCgeLgEYgDM07aPGTM2nKAG8eASADaAawggAXIRIVKAXW2TUmHLwC+fAUNHioASQC2aShHvlBIsdAVQ-hYnZtIjIqCz5QSAdnVwAVcDx5Hn8AoJ1QynCedgB7ciIoay9xaSMGaQ5gKHw5VO0AckEieqhzKDQCaJc3DxtvXmy8gsiIUvopVkrq+TqoRohm1vbOp274yC4gA) displays the issue.

Also, SvelteKit now requires node 16 I believe. Netlify might need to upgrade? I've never used it though so I'm not entirely sure if I'm reading the logs right.

Lastly, I'm going to make this a draft until I get the contentful API keys and can test locally.

**Comment by SirajChokshi** [comment]
Thanks for all the work on this @AndrewLester 🚀 

I think @arpanlaha has access to the Netlify setup, there should be an option to upgrade to Node 16 LTS.

**Comment by arpanlaha** [comment]
I upgraded Node, here's the deploy preview: https://deploy-preview-57--h4iuiuc.netlify.app/ (doesn't display on CI since I went through the netlify thing itself)

**Comment by arpanlaha** [comment]
I added the Ackee env vars to the preview Netlify and we still get a network error (Ackee needs to specify a specific domain which we can't do with preview deploys) but IIUC it isn't blocking the onclick handling now. @AndrewLester could you confirm?

File: .eslintrc.cjs [hunk 1]
```diff
@@ -1,11 +1,16 @@
 module.exports = {
   root: true,
   parser: "@typescript-eslint/parser",
-  extends: ["eslint:recommended", "plugin:@typescript-eslint/recommended"],
+  extends: [
+    "eslint:recommended",
+    "plugin:@typescript-eslint/recommended",
+    "prettier",
```
Comment (arpanlaha) [inline .eslintrc.cjs:7]: I personally like to keep linting and formatting separate - I'm not going to be super hands-on moving forward and no longer "own" this repo so this isn't blocking feedback but I wanted to mention it.

File: .eslintrc.cjs [hunk 2]
```diff
@@ -14,7 +19,7 @@ module.exports = {
   },
   parserOptions: {
     sourceType: "module",
-    ecmaVersion: 2019,
+    ecmaVersion: 2020,
```
Comment (arpanlaha) [inline .eslintrc.cjs:22]: Why this change?

File: src/app.css [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+  --blue-darker: #0d3f6e;
+  --gray-lighter: #ebeef2;
+  --gray-light: #8b9199;
+  --text-dark: #031425;
+
+  --header-width: min(1440px, 90vw);
+  --content-width: min(1200px, 85vw);
+  --long-form-width: min(900px, 80vw);
+
+  --animation: 0.2s ease;
```
Comment (arpanlaha) [inline src/app.css]: I intentially inlined global CSS in app.html - is there a reason you separated it out? I was under the impression that inlining global styles was a good idea to cut down on a network request, but I'm not an expert on this.

File: src/routes/__layout.svelte [hunk 4]
```diff
@@ -1,33 +1,37 @@
...
-    return { props: { semesters, path } };
-  }
+    return { props: { semesters } };
+  };
 </script>
 
 <script lang="ts">
   import * as ackeeTracker from "ackee-tracker";
   import { beforeUpdate } from "svelte";
+  import "../app.css";
```
Comment (arpanlaha) [inline src/routes/__layout.svelte]: If we do go ahead with not inlining global CSS in `app.html`, wouldn't it be more idomatic to add it to the style tag in this file?

File: src/utils/api.ts [hunk 5]
```diff
@@ -25,27 +25,25 @@ export class ContentWrapper {
...
 
-  async get<T>(entity: string, options: any = {}): Promise<T[]> {
+  async get(entity: string, options: any = {}): Promise<any[]> {
     if (this.client === undefined) {
       await this.build();
     }
 
     const [entries, schema] = await Promise.all([
-      this.client.getEntries({
+      this.client!.getEntries({
```
Comment (arpanlaha) [inline src/utils/api.ts]: Looking back at this, instead of doing the `!` assertion it would be better for `this.build` to return `createClient(...)` and change the beginning lines of this function to:

```
if (this.client === undefined) {
      this.client = await this.build();
}
```

From there, TypeScript will be able to figure out that `this.client` exists without the `!` assertion.

File: src/utils/projects.ts [hunk 6]
```diff
@@ -1,15 +1,15 @@
...
   projects: Project[];
-}
+};
 
-export interface ProjectsInfo {
+export type ProjectsInfo = {
   projectMap: Record<string, SemesterProjects>;
   semesters: string[];
-}
+};
```
Comment (arpanlaha) [inline src/utils/projects.ts:12]: Another personal preference for TS - I like to default to `interface` and only use `type` if I'm making use of union types or generics (i.e. `type Status = 'good' | 'bad'` or `type Result = Promise<...>`.

This is also nonblocking feedback

File: src/utils/schema.ts [hunk 7]
```diff
@@ -1,17 +1,21 @@
...
   linkedin: string;
   active: boolean;
   graduationYear: number;
   testimonial?: string;
-}
+};
+
+export type TestimonialMember = Member & {
+  testimonial: string;
+};
```
Comment (arpanlaha) [inline src/utils/schema.ts:18]: Going off of the `interface` vs. `type` comment above, this could be done as `export interface TestimonialMember extends Member {...`

File: src/utils/useViewportAction.ts [hunk 8]
```diff
@@ -12,15 +12,15 @@ function ensureIntersectionObserver() {
   });
 }
 
-export default function viewport(element: HTMLSpanElement) {
+export default function viewport(element: HTMLElement) {
   if (element) {
     ensureIntersectionObserver();
 
-    intersectionObserver.observe(element);
+    intersectionObserver!.observe(element);
```
Comment (arpanlaha) [inline src/utils/useViewportAction.ts]: You could use the same approach as the API client to avoid `!` assertions here

**Review by arpanlaha — COMMENTED** [review]
Thanks for doing this, some feedback that is either relatively small or non-blocking:

File: .eslintrc.cjs (hunk 1 above)
Comment (AndrewLester) [reply .eslintrc.cjs:7]: To be honest, I didn't know what this setting did when I added it. I just copied it from the SvelteKit template. After some Googling, it seems as though placing `prettier` in this file enforces a linting and formatting separation, based on [this docs page](https://prettier.io/docs/en/integrating-with-linters.html) and because this PR adds `eslint-config-prettier` to the package.json.

File: .eslintrc.cjs (hunk 2 above)
Comment (AndrewLester) [reply .eslintrc.cjs:22]: Just took this from the new template: https://github.com/sveltejs/kit/blob/master/packages/create-svelte/shared/%2Beslint%2Bprettier%2Btypescript/.eslintrc.cjs

File: src/app.css (hunk 3 above)
Comment (AndrewLester) [reply src/app.css]: I separated it out because I've seen this pattern a lot for SvelteKit, and it's how [the template](https://sveltekit.new) is setup. It does add a network request, but this can be avoided with [inlineStyleThreshold](https://kit.svelte.dev/docs/configuration#inlinestylethreshold). That said, the file is a bit large to set an inline threshold for in my opinion, so in the future I might try to shrink it by putting some of these styles into components and removing them from here, but for now I'll just revert this back to app.html. I'll still keep the inline style threshold at 1024 bytes for any small files currently.

The only downside to inlining css manually is that Vite's dev server doesn't apply HMR to edits. Also, I think it's a bit easier to read when separated into its own file.

File: src/routes/__layout.svelte (hunk 4 above)
Comment (AndrewLester) [reply src/routes/__layout.svelte]: Addressed in the other comment, but the short answer is it's personal preference I guess? If we put it in the style tag we'd need to add `:global` prefixes to a bunch of the selectors. But, it would be easier to find in there.

File: src/utils/api.ts (hunk 5 above)
Comment (AndrewLester) [reply src/utils/api.ts]: I'll go with this.

File: src/utils/projects.ts (hunk 6 above)
Comment (AndrewLester) [reply src/utils/projects.ts:12]: This is a difficult one because there are tradeoffs to both sides since SvelteKit is a bit weird. With interfaces, another type param must be added whenever these interfaces are returned from an endpoint function. See here: https://github.com/sveltejs/kit/issues/5468#issuecomment-1182231341
Also, the RequestHandler type must be imported from the SvelteKit generated types which isn't as much of an issue.

Using a type, this is not necessary. I personally think it looks cleaner when it's implicit, but I'm not entirely sure if this is enough to justify a change. Here's a TS playground that shows roughly why interface typed objects can't be returned from endpoint functions (without the steps in the comment above): [playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAUgygeQHIDUCGAbArtAvFAZ2ACcBLAOwHMoAfWRJBAIwCsIBjYAbgCgeLgEYgDM07aPGTM2nKAG8eASADaAawggAXIRIVKAXW2TUmHLwC+fAUNHioASQC2aShHvlBIsdAVQ-hYnZtIjIqCz5QSAdnVwAVcDx5Hn8AoJ1QynCedgB7ciIoay9xaSMGaQ5gKHw5VO0AckEieqhzKDQCaJc3DxtvXmy8gsiIUvopVkrq+TqoRohm1vbOp274yC4gA)

File: src/utils/schema.ts (hunk 7 above)
Comment (AndrewLester) [reply src/utils/schema.ts:18]: Will update this to reflect a decision on the above.

File: src/utils/projects.ts (hunk 6 above)
Comment (arpanlaha) [reply src/utils/projects.ts:12]: Ah gotcha, since it's a Svelte issue sgtm

**Review by arpanlaha — APPROVED** [review]
Sweet

**Comment by AndrewLester** [comment]
One last change I made was to separate the layout file into two, `__layout-root.svelte` and `__layout@root.svelte`. The root layout file can be inherited directly by any page to "reset" the layout without missing necessary stuff like ackee (and global CSS in the future). The default layout file that inherits from root has all the actual stuff for page layout like the nav and `<main>`.

**Review by SirajChokshi — APPROVED** [review]
Great work 💯
